src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1430,22 +1430,30 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
 
         // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
-        final boolean useMultiResourceRefreshToken = cachedItem == null || cachedItem.getIsMultiResourceRefreshToken();
-        final RefreshItem refreshItem = getRefreshToken(request, useMultiResourceRefreshToken);
+        final RefreshItem refreshItem = getRefreshToken(request);
         AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
                 && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
-            Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
+            Logger.v(TAG, "Refresh token is available and it will attempt to redeem access token.");
             try {
                 authResult = getTokenWithRefreshTokenAndUpdateCache(request, refreshItem);
             } catch (AuthenticationException authenticationException) {
-                ClientAnalytics.logEvent(
-                        InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED,
-                        new InstrumentationPropertiesBuilder(request, authenticationException).build());
+                addLogEventForRequestFailureWithCause(InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED, request, authenticationException);
                 callbackHandle.onError(authenticationException);
                 return null;
             }
             
+            // retry with MRRT if the refresh token request was done with FRT
+            if (!isAccessTokenReturned(authResult) && refreshItem != null) {
+                try {
+                    authResult = retryWithMRRTIfFRTFail(request, refreshItem);
```

Comment:
Now I'm not sure why do we want to make that sequence: try with FRT and then fallback MRRT
When it would be much cleaner to try MRT first and then if it fails or not fount fallback to FRT.